File: cpp/src/Graph.cpp
#include "../include/Graph.h"
#include <iostream>
#include <algorithm>

Graph::Graph() {
    // Constructor - empty graph
}

Graph::~Graph() {
    // Destructor - cleanup handled by STL containers
}

void Graph::addVertex(const std::string& vertex) {
    if (vertices.find(vertex) == vertices.end()) {
        vertices.insert(vertex);
        adjacencyList[vertex] = std::vector<std::pair<std::string, int>>();
    }
}

void Graph::addEdge(const std::string& from, const std::string& to, int weight) {
    // Add vertices if they don't exist
    addVertex(from);
    addVertex(to);
    
    // Add edge (assuming undirected graph for simplicity)
    // Check if edge already exists
    bool edgeExists = false;
    for (const auto& neighbor : adjacencyList[from]) {
        if (neighbor.first == to) {
            edgeExists = true;
            break;
        }
    }
    
    if (!edgeExists) {
        adjacencyList[from].push_back({to, weight});
        adjacencyList[to].push_back({from, weight});
    }
}

void Graph::removeVertex(const std::string& vertex) {
    if (vertices.find(vertex) == vertices.end()) {
        return; // Vertex doesn't exist
    }
    
    // Remove all edges connected to this vertex
    for (const auto& neighbor : adjacencyList[vertex]) {
        auto& neighbors = adjacencyList[neighbor.first];
        neighbors.erase(
            std::remove_if(neighbors.begin(), neighbors.end(),
                [&vertex](const std::pair<std::string, int>& edge) {
                    return edge.first == vertex;
                }),
            neighbors.end()
        );
    }
    
    // Remove the vertex
    adjacencyList.erase(vertex);
    vertices.erase(vertex);
}

void Graph::removeEdge(const std::string& from, const std::string& to) {
    if (!hasEdge(from, to)) {
        return; // Edge doesn't exist
    }
    
    // Remove edge from both directions (undirected graph)
    auto& fromNeighbors = adjacencyList[from];
    fromNeighbors.erase(
        std::remove_if(fromNeighbors.begin(), fromNeighbors.end(),
            [&to](const std::pair<std::string, int>& edge) {
                return edge.first == to;
            }),
        fromNeighbors.end()
    );
    
    auto& toNeighbors = adjacencyList[to];
    toNeighbors.erase(
        std::remove_if(toNeighbors.begin(), toNeighbors.end(),
            [&from](const std::pair<std::string, int>& edge) {
                return edge.first == from;
            }),
        toNeighbors.end()
    );
}

bool Graph::hasVertex(const std::string& vertex) const {
    return vertices.find(vertex) != vertices.end();
}

bool Graph::hasEdge(const std::string& from, const std::string& to) const {
    if (!hasVertex(from) || !hasVertex(to)) {
        return false;
    }
    
    const auto& neighbors = adjacencyList.at(from);
    for (const auto& neighbor : neighbors) {
        if (neighbor.first == to) {
            return true;
        }
    }
    return false;
}

int Graph::getEdgeWeight(const std::string& from, const std::string& to) const {
    if (!hasEdge(from, to)) {
        return -1; // Edge doesn't exist
    }
    
    const auto& neighbors = adjacencyList.at(from);
    for (const auto& neighbor : neighbors) {
        if (neighbor.first == to) {
            return neighbor.second;
        }
    }
    return -1;
}

std::vector<std::string> Graph::getNeighbors(const std::string& vertex) const {
    std::vector<std::string> neighbors;
    
    if (hasVertex(vertex)) {
        const auto& neighborList = adjacencyList.at(vertex);
        for (const auto& neighbor : neighborList) {
            neighbors.push_back(neighbor.first);
        }
    }
    
    return neighbors;
}

std::vector<std::string> Graph::getVertices() const {
    std::vector<std::string> vertexList;
    for (const auto& vertex : vertices) {
        vertexList.push_back(vertex);
    }
    return vertexList;
}

size_t Graph::getVertexCount() const {
    return vertices.size();
}

size_t Graph::getEdgeCount() const {
    size_t count = 0;
    for (const auto& pair : adjacencyList) {
        count += pair.second.size();
    }
    return count / 2; // Divide by 2 for undirected graph
}

void Graph::displayGraph() const {
    std::cout << "Vertices: ";
    for (const auto& vertex : vertices) {
        std::cout << vertex << " ";
    }
    std::cout << std::endl;
    
    std::cout << "Edges:" << std::endl;
    for (const auto& pair : adjacencyList) {
        std::cout << "  " << pair.first << " -> ";
        for (size_t i = 0; i < pair.second.size(); ++i) {
            std::cout << "(" << pair.second[i].first << ", " << pair.second[i].second << ")";
            if (i < pair.second.size() - 1) {
                std::cout << ", ";
            }
        }
        std::cout << std::endl;
    }
}

void Graph::clear() {
    adjacencyList.clear();
    vertices.clear();
}

const std::unordered_map<std::string, std::vector<std::pair<std::string, int>>>& Graph::getAdjacencyList() const {
    return adjacencyList;
}

File: web/js/datastructures/Graph.js
/**
 * Graph Data Structure
 * Implements adjacency list representation for weighted graphs
 */

class Graph {
    constructor() {
        this.adjacencyList = new Map();
        this.vertices = new Set();
    }

    /**
     * Add a vertex to the graph
     * @param {string} vertex - The vertex identifier
     */
    addVertex(vertex) {
        if (!this.vertices.has(vertex)) {
            this.vertices.add(vertex);
            this.adjacencyList.set(vertex, []);
        }
    }

    /**
     * Add an edge between two vertices with a weight
     * @param {string} from - Source vertex
     * @param {string} to - Destination vertex
     * @param {number} weight - Edge weight (distance/time)
     */
    addEdge(from, to, weight) {
        this.addVertex(from);
        this.addVertex(to);
        
        // Check if edge already exists
        const existingEdge = this.adjacencyList.get(from).find(edge => edge.vertex === to);
        if (!existingEdge) {
            this.adjacencyList.get(from).push({ vertex: to, weight });
            this.adjacencyList.get(to).push({ vertex: from, weight }); // Undirected graph
        }
    }

    /**
     * Remove a vertex from the graph
     * @param {string} vertex - The vertex to remove
     */
    removeVertex(vertex) {
        if (!this.vertices.has(vertex)) return;
        
        // Remove all edges connected to this vertex
        const neighbors = this.adjacencyList.get(vertex);
        neighbors.forEach(neighbor => {
            const neighborList = this.adjacencyList.get(neighbor.vertex);
            const index = neighborList.findIndex(edge => edge.vertex === vertex);
            if (index !== -1) {
                neighborList.splice(index, 1);
            }
        });
        
        this.adjacencyList.delete(vertex);
        this.vertices.delete(vertex);
    }

    /**
     * Remove an edge between two vertices
     * @param {string} from - Source vertex
     * @param {string} to - Destination vertex
     */
    removeEdge(from, to) {
        if (!this.hasEdge(from, to)) return;
        
        // Remove edge from both directions
        const fromList = this.adjacencyList.get(from);
        const toList = this.adjacencyList.get(to);
        
        const fromIndex = fromList.findIndex(edge => edge.vertex === to);
        const toIndex = toList.findIndex(edge => edge.vertex === from);
        
        if (fromIndex !== -1) fromList.splice(fromIndex, 1);
        if (toIndex !== -1) toList.splice(toIndex, 1);
    }

    /**
     * Check if a vertex exists
     * @param {string} vertex - The vertex to check
     * @returns {boolean} True if vertex exists
     */
    hasVertex(vertex) {
        return this.vertices.has(vertex);
    }

    /**
     * Check if an edge exists
     * @param {string} from - Source vertex
     * @param {string} to - Destination vertex
     * @returns {boolean} True if edge exists
     */
    hasEdge(from, to) {
        if (!this.hasVertex(from) || !this.hasVertex(to)) return false;
        
        return this.adjacencyList.get(from).some(edge => edge.vertex === to);
    }

    /**
     * Get the weight of an edge
     * @param {string} from - Source vertex
     * @param {string} to - Destination vertex
     * @returns {number} Edge weight or -1 if not found
     */
    getEdgeWeight(from, to) {
        if (!this.hasEdge(from, to)) return -1;
        
        const edge = this.adjacencyList.get(from).find(edge => edge.vertex === to);
        return edge ? edge.weight : -1;
    }

    /**
     * Get all neighbors of a vertex
     * @param {string} vertex - The vertex
     * @returns {string[]} Array of neighbor vertices
     */
    getNeighbors(vertex) {
        if (!this.hasVertex(vertex)) return [];
        
        return this.adjacencyList.get(vertex).map(edge => edge.vertex);
    }

    /**
     * Get all vertices
     * @returns {string[]} Array of all vertices
     */
    getVertices() {
        return Array.from(this.vertices);
    }

    /**
     * Get the number of vertices
     * @returns {number} Number of vertices
     */
    getVertexCount() {
        return this.vertices.size;
    }

    /**
     * Get the number of edges
     * @returns {number} Number of edgesu
     */
    getEdgeCount() {
        let count = 0;
        for (const [vertex, edges] of this.adjacencyList) {
            count += edges.length;
        }
        return count / 2; // Divide by 2 for undirected graph
    }

    /**
     * Display the graph structure
     */
    displayGraph() {
        console.log('Vertices:', Array.from(this.vertices).join(' '));
        console.log('Edges:');
        for (const [vertex, edges] of this.adjacencyList) {
            const edgeStr = edges.map(edge => `(${edge.vertex}, ${edge.weight})`).join(', ');
            console.log(`  ${vertex} -> ${edgeStr}`);
        }
    }

    /**
     * Clear the graph
     */
    clear() {
        this.adjacencyList.clear();
        this.vertices.clear();
    }

    /**
     * Get adjacency list (for algorithms)
     * @returns {Map} The adjacency list
     */
    getAdjacencyList() {
        return this.adjacencyList;
    }
}
